10up/ElasticPress
# Move default autosuggest selectors definition so they can be filtered.

### Description of the Change

The autosuggest feature is automatically added to these selectors `.ep-autosuggest, input[type="search"], .search-field`. These are defined in the JavaScript file, and cannot be changed by the user. This PR moves the definition of the default selectors into the script localization so that they can be changed by filtering the `defaultSelectors` option with the existing `ep_autosuggest_options` filter.

It also adds a `ep_autosuggest_default_selectors` filter just for the default selectors.

### Alternate Designs

In #2168 @johnbillion suggested a `selectors` option, which could take precedence over the `selector` option if it was present. I opted to just put the default selectors into a `defaultSelectors` option to avoid potential ambiguity between `selector` and `selectors`, and possible confusion due to what would be the absence of an existing `selectors` option in the values that are being filtered.

Another approach could have been to implement the filter in JavaScript, since that's where the original value was, but I opted for putting them inside the PHP filter since that PHP filter was already established as being available to change some selector behaviour.

### Benefits

The most obvious use case would be to stop the plugin adding Autosuggest to all search inputs, by removing the `input[type="search"]` selector using:

```
add_filter(
	'ep_autosuggest_options',
	function($options) {
		$options['defaultSelectors'] = '.ep-autosuggest, .search-field';

		return $options;
	}
);
```
or

```
add_filter(
	'ep_autosuggest_default_selectors',
	function($selectors) {
		$selectors = '.ep-autosuggest, .search-field';

		return $selectors;
	}
);
```

The existence of an additional filter specifically for selectors enables removing default selectors with just:

```
add_filter( 'ep_autosuggest_default_selectors', '__return_empty_string' );
```

### Possible Drawbacks

In the event that a user has used the `ep_autosuggest_options` filter in such a way that does not pass through the `defaultSelectors` option, autosuggest would no longer be added to the default selectors.

### Verification Process

1. Enable Autosuggest and add a search widget. Do not enter anything into the selector setting.
2. Confirm that autosuggest is working.
3. Add code like this to remove the default selectors:
```
add_filter( 'ep_autosuggest_default_selectors', '__return_empty_string' );
```
4. As long as the theme does not add an `ep-autosuggest` class itself, no autosuggest functionality should be present in the search widget.

### Checklist:






### Applicable Issues

Addresses #2168.

### Changelog Entry

Allows changing or removing the default selectors used for autosuggest by filtering the `defaultSelectors` option in the  `ep_autosuggest_options` filter.
